genesis/genesis.go

```diff
@@ -0,0 +1,320 @@
+package genesis
+
+import (
+	"encoding/hex"
+	"fmt"
+	"math/big"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/data/state"
+	"github.com/ElrondNetwork/elrond-go/sharding"
+)
+
+const decodeBase = 10
+
+// Genesis hold data for decoded data from json file
+type Genesis struct {
+	initialBalances []*InitialBalance
+	entireSupply    *big.Int
+}
+
+// NewGenesis creates a new decoded genesis structure from json config file
+func NewGenesis(
+	genesisFilePath string,
+	entireSupply *big.Int,
+) (*Genesis, error) {
+
+	if entireSupply == nil {
+		return nil, ErrNilEntireSupply
+	}
+	if big.NewInt(0).Cmp(entireSupply) >= 0 {
+		return nil, ErrInvalidEntireSupply
+	}
+
+	initialBalances := make([]*InitialBalance, 0)
+	err := core.LoadJsonFile(&initialBalances, genesisFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	genesis := &Genesis{
+		initialBalances: initialBalances,
+		entireSupply:    entireSupply,
+	}
+
+	err = genesis.process()
+	if err != nil {
+		return nil, err
+	}
+
+	return genesis, nil
+}
+
+func (g *Genesis) process() error {
+	totalSupply := big.NewInt(0)
+	for _, initialBalance := range g.initialBalances {
+		err := g.tryParseElement(initialBalance)
+		if err != nil {
+			return err
+		}
+
+		err = g.checkInitialBalance(initialBalance)
+		if err != nil {
+			return err
+		}
+
+		totalSupply.Add(totalSupply, initialBalance.supply)
+	}
+
+	err := g.checkForDuplicates()
+	if err != nil {
+		return err
+	}
+
+	if totalSupply.Cmp(g.entireSupply) != 0 {
+		return fmt.Errorf("%w for entire supply provided %s, computed %s",
+			ErrEntireSupplyMismatch,
+			g.entireSupply.String(),
+			totalSupply.String(),
+		)
+	}
+
+	return nil
+}
+
+func (g *Genesis) tryParseElement(initialBalance *InitialBalance) error {
```

Comment:
Name without `try`, method can return error, so `try` is implied

Reply:
done